src/check.rs

```diff
@@ -94,42 +245,213 @@ fn check_http_url(url: &Url, ctx: &CheckContext) -> Result<(), CheckError> {
         }
     }
 
-    let resp = ureq::head(url.as_str()).call();
-    if resp.synthetic() {
-        Err(CheckError::Http(Box::new(HttpError::Fetch(
-            url.clone(),
-            resp.into_synthetic_error().unwrap(),
-        ))))
-    } else if resp.ok() {
-        Ok(())
+    // only if the URL contains a fragment we need to fetch the body via
+    // GET, otherwise we can use HEAD
+    if url.fragment().is_none() {
+        let resp = ureq::head(url.as_str()).call();
+
+        handle_response(url, resp).map(|_: ureq::Response| ())
     } else {
-        Err(CheckError::Http(Box::new(HttpError::UnexpectedStatus(
-            url.clone(),
-            resp,
-        ))))
+        // the URL might contain a fragment, in that case we need to check if
+        // the fragment exists, this issues a GET request
+        check_http_fragment(url, url.fragment().unwrap())
     }
 }
 
+fn check_http_fragment(url: &Url, fragment: &str) -> Result<(), CheckError> {
+    debug!("Checking fragment {} of URL {}.", fragment, url.as_str());
+
+    let html = || {
+        let mut url = url.clone();
+        url.set_fragment(None);
```

Comment:
Is this necessary? What happens if you don't call `set_fragment`?

Reply:
Yeah, I refactored that wrong, though I do need to fix the caching for HTTP Links, which I think this was doing before the refactor to Link. Cache needs to be keyed to URL without the fragment, otherwise it'll miss for different fragments.